sled-agent/src/instance_manager.rs

```diff
@@ -87,97 +87,121 @@ impl InstanceManager {
-                        // instance's ID. Given that we just replaced said instance using
-                        // the same ID, that would inadvertantly remove our newly created
-                        // instance instead.
-                        // TODO: cleanup source instance properly
-                        std::mem::forget(old_instance);
-                    }
+            } else {
+                info!(&self.inner.log, "new instance");
+                let instance_log = self.inner.log.new(o!());
+                let ticket =
+                    InstanceTicket::new(instance_id, self.inner.clone());
+                let instance = Instance::new(
+                    instance_log,
+                    instance_id,
+                    ticket,
+                    initial_hardware,
+                    self.inner.vnic_allocator.clone(),
+                    self.inner.port_manager.clone(),
+                    self.inner.lazy_nexus_client.clone(),
+                )?;
+                let instance_clone = instance.clone();
+                let _old = instances
+                    .insert(instance_id, (requested_propolis_id, instance));
+                assert!(_old.is_none());
+                instance_clone
+            }
+        };
 
-                    let ticket = Some(InstanceTicket::new(
-                        instance_id,
-                        self.inner.clone(),
-                    ));
-                    (instance_clone, ticket)
+        Ok(instance.current_state().await)
+    }
+
+    /// Idempotently attempts to drive the supplied instance into the supplied
+    /// runtime state.
+    pub async fn ensure_state(
+        &self,
+        instance_id: Uuid,
+        target: InstanceStateRequested,
+    ) -> Result<InstancePutStateResponse, Error> {
+        let instance = {
+            let instances = self.inner.instances.lock().unwrap();
+            let instance = instances.get(&instance_id);
+
+            if let Some((_, instance)) = instance {
+                instance.clone()
+            } else {
+                match target {
+                    // If the instance isn't registered, then by definition it
+                    // isn't running here. Allow requests to stop or destroy the
+                    // instance to succeed to provide idempotency. This has to
+                    // be handled here (that is, on the "instance not found"
+                    // path) to handle the case where a stop request arrived,
+                    // Propolis handled it, sled agent unregistered the
+                    // instance, and only then did a second stop request
+                    // arrive.
+                    InstanceStateRequested::Stopped
+                    | InstanceStateRequested::Destroyed => {
+                        return Ok(InstancePutStateResponse {
+                            updated_runtime: None,
+                        });
+                    }
```

Comment:
(Not a request, more just thinking out loud)

This special casing makes me wonder if the sled agent should expose a separate API for fully "destroying the instance" instead of relying on particular state transitions to pull that off for us -- similar to how we divided "creation" away from "starting".

Reply:
Acknowledged. I think that makes sense, but will leave things as they are in this particular PR.